pypeit/core/wavecal/fitting.py

```diff
@@ -9,10 +9,98 @@
 from pypeit.core.wavecal import qa
 from pypeit import msgs
 
-from pypeit import debugger
 
+def fit_slit(spec, patt_dict, tcent, line_lists, outroot=None, slittxt="Slit", thar=False,match_toler=3.0,
+             func='legendre', n_first=2,sigrej_first=2.0,n_final=4,sigrej_final=3.0,verbose=False):
+
+    """ Perform a fit to the wavelength solution. Wrapper for iterative fitting code.
+
+    Parameters
+    ----------
+    spec : ndarray
+      arc spectrum
+    patt_dict : dict
+      dictionary of patterns
+    tcent: ndarray
+      List of the detections in this slit to be fit using the patt_dict
+    outroot : str
+      root directory to save QA
+
+    Optional Parameters
+    -------------------
+    outroot: str
+      Path for QA file.
+    slittxt : str
+      Label used for QA
+    thar: bool, default = False
+      True if this is a ThAr fit
+    match_toler: float, default = 3.0
+      Matching tolerance when searching for new lines. This is the difference in pixels between the wavlength assigned to
+      an arc line by an iteration of the wavelength solution to the wavelength in the line list.
+    func: str, default = 'legendre'
+      Name of function used for the wavelength solution
+    n_first: int, default = 2
+      Order of first guess to the wavelength solution.
+    sigrej_first: float, default = 2.0
+      Number of sigma for rejection for the first guess to the wavelength solution.
+    n_final: int, default = 4
+      Order of the final wavelength solution fit
+    sigrej_final: float, default = 3.0
+      Number of sigma for rejection for the final fit to the wavelength solution.
+    verbose : bool
+      If True, print out more information.
+    plot_fil:
+      Filename for plotting some QA?
+
+    Returns
+    -------
+    final_fit : dict
+      A dictionary containing all of the information about the fit
+    """
+
+    # Check that patt_dict and tcent refer to each other
+    if patt_dict['mask'].shape != tcent.shape:
+        msgs.error('patt_dict and tcent do not refer to each other. Something is very wrong')
+
+    # Perform final fit to the line IDs
+    if thar:
+        NIST_lines = (line_lists['NIST'] > 0) & (np.char.find(line_lists['Source'].data, 'MURPHY') >= 0)
+    else:
+        NIST_lines = line_lists['NIST'] > 0
+    ifit = np.where(patt_dict['mask'])[0]
+
+    if outroot is not None:
+        plot_fil = outroot + slittxt + '_fit.pdf'
+    else:
+        plot_fil = None
+
+    # TODO Profx maybe you can add a comment on what this is doing. Why do we have use_unknowns=True only to purge them later??
```

Comment:
Should be because the UNKWNS are used to find IDs but then have imprecise (i.e. non-NIST) wavelengths so are not included in the fit.

Reply:
Okay my suggestion is that we keep separate line lists as arguments and only pass in the ones we need. By this stage of the routine, the IDs are finished, and so this routine only cares about fitting. This unknowns syntax here makes a simple code look much more complicated.